addons/web_editor/static/src/js/editor/snippets.options.js

```diff
@@ -6964,6 +6964,19 @@ registry.BackgroundShape = SnippetOptionWidget.extend({
                 .prepend($(`<we-colorpicker data-color="true" data-color-name="${colorName}"></we-colorpicker>`)[0]);
         });
```

Comment:
Just an idea I had: 
Since you do all this to get the shape rules "only" in order to replace the background image URLs by the color-updated ones, and since the only difference between the URLs is the color parameters (ex: `?c3=%23F6F6F6`), have you considered simply updating the URLs parameters ? 
You could get the different colors with 
```js
window.getComputedStyle(this.$target[0]).getPropertyValue("--o-cc1-bg")
```
and then replace the color codes in the URL by matching the strings 

Not sure if it would be better or not, just a suggestion :wink: After all, the "getting" of the shape rules is only made once, so I don't think it is really a bad thing to browse all the rules :thinking:

Reply:
Thanks for the suggestion.
With the current solution we only need to replace a property with a string.
Doing the palette replacement manually involves a lot more operations: matching with a regex, replacing parts accoring to some conditions... and it still needs to be done for every for each shape image.
Also in terms of complexity it is _O(number of shapes)_ vs _O(number of shapes * number of colors in shape)_ (even if the number of colors is capped at 5) at every custom XML rendering, while the rule inventory is done only once.
Therefore I'd rather keep the current solution.

Reply:
Just found an issue: it does not work if a single preset color is changed, but your suggestion would work.
Investigation in progress.

Reply:
Ok, here is the conclusion of the investigation.
Steps to reproduce:
- have a block with a shape
- select it
- go to themes
- open the primary or secondary color picker
- pick a color but do not close the color picker
- click on the BLOCKS tab then click on the snippet, or click on the snippet directly

=> the shape dropdown was not updated with the color change

There were two problems:
1) `_onReloadBundles` only impacted the editors if the theme tab remained selected => now, it asks the editors to redraw their UI if the theme tab is not selected anymore
2) the shape URLs were only refreshed upon renewal of the option - which does not happen in this case => the shape URLs cache is now invalidated if the palette changed